types/state/stateTxn.go

```diff
@@ -0,0 +1,334 @@
+		return tx.state.bpt.Bpt.Root.Hash[:], 0, nil
+	}
+
+	tx.state.blockIndex = blockHeight
+	// TODO MainIndex and PendingIndex?
+	tx.AddStateEntry((*types.Bytes32)(&blockIndexKey), new(types.Bytes32), &Object{Entry: common.Int64Bytes(blockHeight)})
+
+	group := new(sync.WaitGroup)
+	group.Add(1)
+	group.Add(len(tx.state.updates))
+
+	mutex := new(sync.Mutex)
+	//to try the multi-threading add "go" in front of the next line
+	err := tx.writeTxs(mutex, group)
+	if err != nil {
+		return tx.state.bpt.Bpt.Root.Hash[:], 0, nil
+	}
+
+	// Create an ordered list of chain IDs that need updating. The iteration
+	// order of maps in Go is random. Randomly ordering database writes is bad,
+	// because that leads to consensus errors between nodes, since each node
+	// will have a different random order. So we need updates to have some
+	// consistent order, regardless of what it is.
+	updateOrder := make([]types.Bytes32, 0, len(tx.state.updates))
+	for id := range tx.state.updates {
+		updateOrder = append(updateOrder, id)
+	}
+	sort.Slice(updateOrder, func(i, j int) bool {
+		return bytes.Compare(updateOrder[i][:], updateOrder[j][:]) < 0
+	})
+
+	for _, chainId := range updateOrder {
+		tx.state.mm.SetChainID(chainId[:])
+		tx.state.writeChainState(group, mutex, tx.state.mm, chainId)
+
+		//TODO: figure out how to do this with new way state is derived
+		//if len(currentState.pendingTx) != 0 {
+		//	mdRoot := v.PendingChain.MS.GetMDRoot()
+		//	if mdRoot == nil {
+		//		//shouldn't get here, but will reject if I do
+		//		panic(fmt.Sprintf("shouldn't get here on writeState() on chain id %X obtaining merkle state", chainId))
+		//	}
+		//	//todo:  Determine how we purge pending tx's after 2 weeks.
+		//	s.bpt.Bpt.Insert(chainId, *mdRoot)
+		//}
+	}
+
+	// Process pending writes
+	writeOrder := make([]storage.Key, 0, len(tx.state.writes))
+	for k := range tx.state.writes {
+		writeOrder = append(writeOrder, k)
+	}
+	sort.Slice(writeOrder, func(i, j int) bool {
+		return bytes.Compare(writeOrder[i][:], writeOrder[j][:]) < 0
+	})
+	for _, k := range writeOrder {
+		tx.state.GetDB().Key(k).PutBatch(tx.state.writes[k])
+	}
+	// The compiler optimizes this into a constant-time operation
+	for k := range tx.state.writes {
+		delete(tx.state.writes, k)
+	}
+
+	group.Wait()
+
+	tx.state.bpt.Bpt.Update()
+
+	//reset out block update buffer to get ready for the next round
+	tx.state.sync.Add(1)
+	//to enable threaded batch writes, put go in front of next line.
+	tx.state.writeBatches()
+
+	tx.state.updates = make(map[types.Bytes32]*blockUpdates)
+
+	//return the state of the BPT for the state of the block
+	rh := types.Bytes(tx.state.RootHash()).AsBytes32()
+	tx.state.logInfo("WriteStates", "height", blockHeight, "hash", logging.AsHex(rh))
+	return tx.state.RootHash(), currentStateCount, nil
+	panic("TODO")
+}
+
+func (tx *DBTransaction) writeChainState(group *sync.WaitGroup, mutex *sync.Mutex, mm *managed.MerkleManager, chainId types.Bytes32) {
```

Comment:
Instead of adding a copy of `writeChainState`, change the receiver of the original.